java/src/main/native/src/TableJni.cpp

```diff
@@ -656,6 +658,69 @@ std::vector<cudf::null_order> resolve_null_precedence(JNIEnv *env, jbooleanArray
 
 namespace {
 
+int set_column_metadata(cudf::io::column_in_metadata &column_metadata,
+                         std::vector<std::string> &col_names, 
+                         cudf::jni::native_jbooleanArray &nullability,
+                         cudf::jni::native_jbooleanArray &isInt96,
+                         cudf::jni::native_jintArray &precisions,
+                         cudf::jni::native_jintArray &children, int num_children, int read_index) {
+  int write_index = 0;
+  for (int i = 0; i < num_children; i++, write_index++) {
+    cudf::io::column_in_metadata child;
+    child.set_name(col_names[read_index])
+        .set_decimal_precision(precisions[read_index])
+        .set_int96_timestamps(isInt96[read_index])
+        .set_nullability(nullability[read_index]);
+    column_metadata.add_child(child);
+    if (children[read_index] > 0) {
+      read_index = set_column_metadata(column_metadata.child(write_index), col_names, nullability,
+                                       isInt96, precisions, children, children[read_index], read_index + 1);
+    } else {
+      read_index++;
+    }
```

Comment:
The point of passing the number of children as an argument is to simplify this logic as well, e.g.:
```suggestion
    int child_num_children = children[read_index++];
    if (child_num_children > 0) {
      read_index = set_column_metadata(column_metadata.child(write_index), col_names, nullability,
                                       isInt96, precisions, children, child_num_children, read_index);
    }
```